### ISSUE TITLE
Dask: ValueError: Length mismatch: Expected axis has 2 elements, new values have 238 elements

### ISSUE THREAD

**Issue by adrianodemarino** [dask/dask#3225]
I'm using dask library to compute some operation not able using pandas. 

After Creating a list of dask dataframe, like this

Mylist = [ddf1,ddf2,ddf3,ddf4]
Each ddf in list is very big to compute() ! 

 I concat them into single one using:
`concat = dd.multi.concat(Mylist,interleave_partitions=True)`

Inside that I know that are present different NaN values. I would like to remove them using:
`concatFilledNA = concat.fillna(0, axis=1)`
This command is performed but, when I Try to iterate over the big concatenated using:
`concatFilledNA.iterrows()` dask data frame concatFilledNA i read this error : 

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-24-374d1c2f9052> in <module>()
----> 1 for index, columns in concatnull.iterrows():                #<--- TOTAL_SHARED_RISUMI
      2                 print(len(columns.tolist()))
      3
/home/adriano/.local/lib/python2.7/site-packages/dask/dataframe/core.pyc in iterrows(self)
   2645     def iterrows(self):
   2646         for i in range(self.npartitions):
-> 2647             df = self.get_partition(i).compute()
   2648             for row in df.iterrows():
   2649                 yield row
/home/adriano/.local/lib/python2.7/site-packages/dask/base.pyc in compute(self, **kwargs)
    141         dask.base.compute
    142         """
--> 143         (result,) = compute(self, traverse=False, **kwargs)
    144         return result
    145
/home/adriano/.local/lib/python2.7/site-packages/dask/base.pyc in compute(*args, **kwargs)
    390     postcomputes = [a.__dask_postcompute__() if is_dask_collection(a)
    391                     else (None, a) for a in args]
--> 392     results = get(dsk, keys, **kwargs)
    393     results_iter = iter(results)
    394     return tuple(a if f is None else f(next(results_iter), *a)
/home/adriano/.local/lib/python2.7/site-packages/dask/threaded.pyc in get(dsk, result, cache, num_workers, **kwargs)
     73     results = get_async(pool.apply_async, len(pool._pool), dsk, result,
     74                         cache=cache, get_id=_thread_get_id,
---> 75                         pack_exception=pack_exception, **kwargs)
     76
     77     # Cleanup pools associated to dead threads
/home/adriano/.local/lib/python2.7/site-packages/dask/local.pyc in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    519                         _execute_task(task, data)  # Re-execute locally
    520                     else:
--> 521                         raise_exception(exc, tb)
    522                 res, worker_id = loads(res_info)
    523                 state['cache'][key] = res
/home/adriano/.local/lib/python2.7/site-packages/dask/local.pyc in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    288     try:
    289         task, data = loads(task_info)
--> 290         result = _execute_task(task, data)
    291         id = get_id()
    292         result = dumps((result, id))

/home/adriano/.local/lib/python2.7/site-packages/dask/local.pyc in _execute_task(arg, cache, dsk)
    269         func, args = arg[0], arg[1:]
    270         args2 = [_execute_task(a, cache) for a in args]
--> 271         return func(*args2)
    272     elif not ishashable(arg):
    273         return arg
/home/adriano/.local/lib/python2.7/site-packages/dask/dataframe/core.pyc in apply_and_enforce(func, args, kwargs, meta)
   3354             return meta
   3355         c = meta.columns if isinstance(df, pd.DataFrame) else meta.name
-> 3356         return _rename(c, df)
   3357     return df
   3358
/home/adriano/.local/lib/python2.7/site-packages/dask/dataframe/core.pyc in _rename(columns, df)
   3391         # deep=False doesn't doesn't copy any data/indices, so this is cheap
   3392         df = df.copy(deep=False)
-> 3393         df.columns = columns
   3394         return df
   3395     elif isinstance(df, (pd.Series, pd.Index)):
/home/adriano/.local/lib/python2.7/site-packages/pandas/core/generic.pyc in __setattr__(self, name, value)
   3625         try:
   3626             object.__getattribute__(self, name)
-> 3627             return object.__setattr__(self, name, value)
   3628         except AttributeError:
   3629             pass
/home/adriano/.local/lib/python2.7/site-packages/pandas/_libs/properties.so in pandas._libs.properties.AxisProperty.__set__()
/home/adriano/.local/lib/python2.7/site-packages/pandas/core/generic.pyc in _set_axis(self, axis, labels)
    557
    558     def _set_axis(self, axis, labels):
--> 559         self._data.set_axis(axis, labels)
    560         self._clear_item_cache()
    561
/home/adriano/.local/lib/python2.7/site-packages/pandas/core/internals.pyc in set_axis(self, axis, new_labels)
   3072             raise ValueError('Length mismatch: Expected axis has %d elements, '
   3073                              'new values have %d elements' %
-> 3074                              (old_len, new_len))
   3075
   3076         self.axes[axis] = new_labels
ValueError: Length mismatch: Expected axis has 2 elements, new values have 238 elements

Instead if I didn't apply the function fillna works!

Can anyone help me to fillna value inside my dask data frame ?

**Comment by mrocklin** [comment]
Here is my attempt to reproduce your error:

```python
import pandas as pd
df = pd.DataFrame({'x': [1, 2, None, 4, 5]})

import dask.dataframe as dd
ddf = dd.from_pandas(df, npartitions=3)
ddf = dd.concat([ddf] * 4, interleave_partitions=True)
ddf = ddf.fillna(0, axis=1)

list(ddf.iterrows())
```

Unfortunately everything works fine.  Can I ask you to help by providing a simple example that reproduces the exception that you've seen?

**Comment by adrianodemarino** [comment]
Sure!,
First of all, thank you for your fast answer. 

**data At the beginning**: 
association_ID  genomic_coordinates     shearsite       randomBC        seq_count
LAM-LTR_Block_L_11_LTR21.LC66_BR2_TR1   chr11_64537168_-        19      AAGAAAGCAATT    32
LAM-LTR_Block_L_11_LTR21.LC66_BR2_TR1   chr11_64537168_-        19      CGGAGCGTATTA    1
LAM-LTR_Block_L_11_LTR21.LC66_BR2_TR1   chr11_64537168_-        19      CGTAGCACCAAA    1
LAM-LTR_Block_L_11_LTR21.LC66_BR2_TR1   chr11_64537168_-        19      TCGCGTCCCGCA    4
LAM-LTR_Block_L_11_LTR21.LC66_BR2_TR1   chr11_64537168_-        21      AAGGAAGTATGA    1
LAM-LTR_Block_L_11_LTR21.LC66_BR2_TR1   chr11_64537168_-        21      AGCATGTCTACT    22
LAM-LTR_Block_L_11_LTR21.LC66_BR2_TR1   chr11_64537168_-        21      AGGACAGCTCGG    22
LAM-LTR_Block_L_11_LTR21.LC66_BR2_TR1   chr11_64537168_-        21      ATAGGATTCAGA    1
LAM-LTR_Block_L_11_LTR21.LC66_BR2_TR1   chr11_64537168_-        21      CAAGCGGTGGCG    21
LAM-LTR_Block_L_11_LTR21.LC66_BR2_TR1   chr11_64537168_-        21      CCCTGCATAGGG    12
...

This is all the interest part of my code:

##load library
from datetime import date
import pandas as pd
import numpy as np
import os, sys, argparse, time, multiprocessing, math, itertools
from multiprocessing import Queue,Process
import dask.dataframe as dd
####################################


1) **Function** to reduce the memory usage of pandas_obj
```
def reduce_memory_df(pandas_obj):
    
    converted_obj = pd.DataFrame()
    np16 = ['shearsite']
    np32 = ['locus','seq_count']
    category = ['chr','strand','association_ID']
    
    uint16 = 2 ** 16

    for col in pandas_obj.columns:
        if col in np32 and int(pandas_obj.loc[:,col].values.max()) > (2 ** 16):
            converted_obj.loc[:,col] = pandas_obj[col].astype(np.uint32)
        elif col in np32 and int(pandas_obj.loc[:,col].values.max()) <= (2 ** 16):
            converted_obj.loc[:,col] = pandas_obj[col].astype(np.uint16)
        elif col in category:
            converted_obj.loc[:,col] = pandas_obj[col].astype('category')
        elif col in np16:
            converted_obj.loc[:,col] = pandas_obj[col].astype(np.uint16)
        else:
            converted_obj.loc[:,col] = pandas_obj[col]

    return converted_obj
```

2. **Load data** and split it into different pieces by chromosome: This function return a list of pandas_obj.

```
def groups(inputc,gzip=gzip):
    '''
    This function repeat the behaivor of th pivot_table but using groupby() and unstuck()
    '''
    dtype = {'genomic_coordinates' : 'category', 'association_ID' : 'category','shearsite': np.uint32, 'seq_count' : np.uint32}
    chunksize = 10 ** 6
    if gzip != None:
        any_df = pd.read_csv(inputc, sep='\t', dtype=dtype, compression='gzip')
    else:
        any_df = pd.read_csv(inputc, sep='\t', dtype=dtype)
    if 'Unnamed: 0' in any_df.columns:
        any_df.drop('Unnamed: 0', axis=1, inplace=True)
    any_df[['chr','locus','strand']] = (any_df.genomic_coordinates.str.split('_', expand= True ))
    any_df = reduce_memory_df(any_df)
    gb = any_df.groupby('chr') #hard to compute() but ok because use only one key and use category data.
    gb_list = []
    for _chr_ in gb:
        gb_list.append(_chr_[1])
    return gb_list
```

3) **transform and concatenate in parallel** each pandas obj in dask,  pivot it and concatenate the list gained.

```
def ppivot_RISUMI(gb):
    
    def worker(gb,out_q):
        """ The worker function, invoked in a process. 'gb' is a
            list of dataframe to pivot. The results are placed in
            a dataframe that's pushed to a queue.
        """    
        gb['RISUMI'] = gb[['genomic_coordinates','shearsite','randomBC']].apply(lambda x: "_".join((x[0], str(x[1]), x[2])), axis=1)
        gb.drop(['genomic_coordinates','shearsite','randomBC','chr','locus','strand'], axis=1, inplace=True)
        ddf = dd.from_pandas(gb,npartitions=5) #convert to pandas structure
        dd_pivot = (ddf.pivot_table(values="seq_count", index="RISUMI", columns="association_ID", aggfunc='sum'))#.astype(np.dtype(np.uint32)) #shape the form & convert datatype
        out_q.put(dd_pivot)
    
    #Each process will get 'chunksize' gb and a queue to put his out
    #Dataframe into
    out_q = Queue()
    procs = []

    # Collect all results into a single result pd.DataFrame(). We know how many pd.DataFrame()
    # with results to expect.
    resultdata = []
    npools = len(gb)
    for i in range(npools):
        p = multiprocessing.Process(target=worker, args=([gb[i],out_q])) 
        procs.append(p)
        p.start()
    for i in range(npools):
        a = out_q.get()
        resultdata.append(a)
    for p in procs:
        p.join()

    concat = dd.multi.concat(resultdata,interleave_partitions=True)
    
    return concat
```
**Data like appear if I compute** `concat.compute()` <-- but I can't perform this with my data

RISUMI  LAM-LTR_Block_L_11_LTR21.LC66_BR2_TR1        LAM-LTR_Block_L_12_LTR21.LC66_BR2_TR1        LAM-LTR_Block_L_11_LTR11.LC36_BR2_TR1
chr10_100200594_-_345_CACGGCTCGAAA      0       0       0   ...
chr10_100200594_-_346_AACGAACTATAT      0       0       0   ...
chr10_100200594_-_346_AACGATACAGCT      0       0       0   ...
...

The final **concat dask obj** if I try to compute it, I will get memory error. Is too big!
If I try to fillna values I will get the error above cited after `iterrows() `.

I'm just trying to get a way to analyse these data without get memory error, for this reason I'm using dask. 

I have 256gb of RAM

**Comment by mrocklin** [comment]
Can i ask you to try to pare things down to a minimal example?  Preferably one that I can copy and paste easily, like the example in my earlier comment.  See https://stackoverflow.com/help/mcve for guidelines

**Comment by adrianodemarino** [comment]
#
[testcase.csv.gz](https://github.com/dask/dask/files/1767369/testcase.csv.gz)
Use this file for example.
And try this code in ipython
```
from datetime import date
import pandas as pd
import numpy as np
import os, sys, argparse, time, multiprocessing, math, itertools
from multiprocessing import Queue,Process
import dask.dataframe as dd

def reduce_memory_df(pandas_obj):
    
    converted_obj = pd.DataFrame()
    np16 = ['shearsite']
    np32 = ['locus','seq_count']
    category = ['chr','strand','association_ID']
    
    uint16 = 2 ** 16

    for col in pandas_obj.columns:
        if col in np32 and int(pandas_obj.loc[:,col].values.max()) > (2 ** 16):
            converted_obj.loc[:,col] = pandas_obj[col].astype(np.uint32)
        elif col in np32 and int(pandas_obj.loc[:,col].values.max()) <= (2 ** 16):
            converted_obj.loc[:,col] = pandas_obj[col].astype(np.uint16)
        elif col in category:
            converted_obj.loc[:,col] = pandas_obj[col].astype('category')
        elif col in np16:
            converted_obj.loc[:,col] = pandas_obj[col].astype(np.uint16)
        else:
            converted_obj.loc[:,col] = pandas_obj[col]

    return converted_obj

def groups(inputc,gzip=gzip):
    '''
    This function repeat the behaivor of th pivot_table but using groupby() and unstuck()
    '''
    dtype = {'genomic_coordinates' : 'category', 'association_ID' : 'category','shearsite': np.uint32, 'seq_count' : np.uint32}
    chunksize = 10 ** 6
    if gzip != None:
        any_df = pd.read_csv(inputc, sep='\t', dtype=dtype, compression='gzip')
    else:
        any_df = pd.read_csv(inputc, sep='\t', dtype=dtype)
    if 'Unnamed: 0' in any_df.columns:
        any_df.drop('Unnamed: 0', axis=1, inplace=True)
    any_df[['chr','locus','strand']] = (any_df.genomic_coordinates.str.split('_', expand= True ))
    any_df = reduce_memory_df(any_df)
    gb = any_df.groupby('chr') #hard to compute() but ok because use only one key and use category data.
    gb_list = []
    for _chr_ in gb:
        gb_list.append(_chr_[1])
    return gb_list

def ppivot_RISUMI(gb):
    
    def worker(gb,out_q):
        """ The worker function, invoked in a process. 'gb' is a
            list of dataframe to pivot. The results are placed in
            a dataframe that's pushed to a queue.
        """    
        gb['RISUMI'] = gb[['genomic_coordinates','shearsite','randomBC']].apply(lambda x: "_".join((x[0], str(x[1]), x[2])), axis=1)
        gb.drop(['genomic_coordinates','shearsite','randomBC','chr','locus','strand'], axis=1, inplace=True)
        ddf = dd.from_pandas(gb,npartitions=5) #convert to pandas structure
        dd_pivot = (ddf.pivot_table(values="seq_count", index="RISUMI", columns="association_ID", aggfunc='sum'))#.astype(np.dtype(np.uint32)) #shape the form & convert datatype
        out_q.put(dd_pivot)
    
    # Each process will get 'chunksize' gb and a queue to put his out
    # Dataframe into
    out_q = Queue()
    procs = []

    # Collect all results into a single result pd.DataFrame(). We know how many pd.DataFrame()
    # with results to expect.
    resultdata = []
    npools = len(gb)
    for i in range(npools):
        p = multiprocessing.Process(target=worker, args=([gb[i],out_q])) 
        procs.append(p)
        p.start()
    for i in range(npools):
        a = out_q.get()
        resultdata.append(a)
    for p in procs:
        p.join()

    concat = dd.multi.concat(resultdata,interleave_partitions=True)
    
    return concat

gzip = 'gzip'
gb = groups('testcase.csv.gz')
concat = ppivot_RISUMI(gb)
concatFilled = concat.fillna(0,axis=1)

list(concatFilled.iterrows())
```

**Comment by mrocklin** [comment]
Sorry to be repetitive here, but can I ask you to simplify this more?  Ideally you keep removing things from your problem that can be removed until you've isolated the one thing that causes it.  Right now there is a lot of stuff going on.  The use of gzip compression, multiprocessing, astype operations, fillna, etc..

It is unlikley that a maintainer will spend their time to review this problem as is.

**Comment by adrianodemarino** [comment]
ok sorry, I will try to do my best to simplify as more as possible.

**Comment by mrocklin** [comment]
This (along with a few thousand other experiences) inspired me to write a tiny blogpost: http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

**Comment by adrianodemarino** [comment]
After reading your blog post, I got it how to reproduce error example. Sorry for before, but it's my first time.

**Comment by mrocklin** [comment]
It's not a problem.  I still have trouble with this myself :)

**Comment by citynorman** [comment]
You can use `d6tstack` library to deal with inconsistent columns across input files. It will make all your input data consistent, you can keep all or only common columns with support for out-of-core. For details see https://github.com/d6t/d6tstack/blob/master/examples-dask.ipynb

**Comment by tslmy** [comment]
Similar problem here, and manual spawning works for me. I'd rather not manually spin up 20 workers manually though...
